Source/ParleyTests/MessagesManagerTests.swift

```diff
@@ -0,0 +1,274 @@
+        XCTAssertEqual(messagesManager.messages.count, 2, "There should now be two one message in the messages array.")
+        
+        let secondInsertionIndexPaths = messagesManager.add(message)
+        XCTAssertTrue(secondInsertionIndexPaths.isEmpty, "Array should be empty because it should not insert the same message twice.")
+        XCTAssertEqual(messagesManager.messages.count, 2, "Array of messages should remain the same.")
+    }
+    
+    func testTypingMessage_ShouldBeLastMessage_WhenAddingTheTypingMessage() {
+        addInfoMessageAndFirstUserMessage()
+        
+        let indexPaths = messagesManager.addTypingMessage()
+        XCTAssertEqual(indexPaths.count, 1, "Adding Typing indicator should only add a single IndexPath.")
+        XCTAssertEqual(indexPaths[0].row, 3, "Typing message should be the fourth index.")
+        XCTAssertEqual(indexPaths[0].section, 0, "Typing message should be the fourth index.")
+        
+        XCTAssertEqual(messagesManager.messages.last!.type, .agentTyping, "Last message should be of type agent typing.")
+    }
+    
+    func testAddTypingIndicator_ShouldBeIgnored_WhenTryingToAddItTwice() {
+        addInfoMessageAndFirstUserMessage()
+        
+        let indexPathsFirstTypingMessage = messagesManager.addTypingMessage()
+        XCTAssertEqual(indexPathsFirstTypingMessage.count, 1)
+        
+        let indexPathsSecondTypingMessage = messagesManager.addTypingMessage()
+        XCTAssertTrue(indexPathsSecondTypingMessage.isEmpty)
+    }
+    
+    func testRemoveTypingIndicator_ShouldReturnNil_WhenTryingToRemoveANonExistentTypingIndicator() {
+        addInfoMessageAndFirstUserMessage()
+        
+        XCTAssertNotEqual(messagesManager.messages.last!.type, .agentTyping, "There shouldn't be a typing message present.")
+        XCTAssertNil(messagesManager.removeTypingMessage(), "Trying to remove a typing message without there being one should result in nil being returned.")
+    }
+    
+    func testTypingIndicator_ShouldBeGone_WhenAddingAndRemvovingTheTypingIndicator() {
+        addInfoMessageAndFirstUserMessage()
+        
+        _ = messagesManager.addTypingMessage()
+        
+        let removedTyingMessageIndexPaths = messagesManager.removeTypingMessage()
+        
+        XCTAssertNotEqual(messagesManager.messages.last!.type, .agentTyping, "Last message should not be a agent typing kind.")
+        XCTAssertNotNil(removedTyingMessageIndexPaths, "Should not be nil because it should contain one IndexPath to remove.")
+        XCTAssertEqual(removedTyingMessageIndexPaths!.count, 1, "Only one element should have been added.")
+        XCTAssertEqual(removedTyingMessageIndexPaths![0].row, 3, "Should remove at row 3.")
+        XCTAssertEqual(removedTyingMessageIndexPaths![0].section, 0, "Section index should always be 0.")
+    }
+    
+    func testAddUserMessage_ShouldRetainCorrectOrder_WhenAgentIsTying() {
```

Comment:
This would be more a `testUserMessage_ShouldBeAddedAfterTyping_WhenAgentIsTyping`